### ISSUE TITLE
Android inputmask looses the first character when i start typing

### ISSUE THREAD

**Issue by daanporon** [RobinHerbots/Inputmask#254]
Hi, i'm experiencing some problems on an android device when implementing the jquery.inputmask plugin. Whenever i click on a text-field which has the inputmask enabled the cursor is located at the end of the field. When i directly start typing then it moves to the front and i loose the first character i typed. 

I tried debugging this myself but i saw there are a lot of checks for android devices that do do something specific regarding the selectionStart and selectionEnd property and i don't know the history why these checks are there. For example in the caret function a timeout has been set for 10 milliseconds before the selectionStart and selectionEnd is set. My issue is probably related to this. I know the timeout is only 10 milliseconds ... but maybe if your on an android tablet which isn't that quick the 10 milliseconds take longer because he's still processing other javascript or so ... 

Basically the inputmask is ready to use when your cursor is moved to the front of the inputfield. Isn't it possible to buffer the keys that are typed until then and then replay them or something similar.

I tested in version 4.1.2

Later today i can provide a link where the jquery.inputmask plugin is implemented.

kind regards,
Daan

**Comment by RobinHerbots** [comment]
@daanporon, 

I have done a refactoring of the plugin to better handle multiple mask.  The plugins needs some more stabilization at the moment,   

I haven't tested the new code on android yet.  I will do some testing on android soon.

Did you test on the standard browser or another mobile one?

best regards, 

Robin

**Comment by daanporon** [comment]
We tested in the default android browser. The page where we integrated the inputmask can be found here: https://www.bolero.be/nl/word-client/uw-gegevens
But for now we disabled the feature on android and did a fallback to a placeholder.

**Comment by Dalmadog** [comment]
Is this occuring in a modal dialog ?

If so, I think this may be related to :
- the double keystroke hack for Safari in modal dialogs
- Chrome using the same js engine as Safari (therefor same useragent maybe ?)
- Android using Chrome ?

I remember trying to solve the double keystroke bug for Safari but not having it in Chrome. Maybe the Android hack is not correct ? (maybe)

**Comment by roderik** [comment]
@Dalmadog no, this does not happen in modal, just a plain page.

**Comment by RobinHerbots** [comment]
@daanporon, 

Can you specify the mask with problems or do you experience this with all masks?  Which version of the inputmask are you using?

I tested on android 4.1.2 & 4.2.2 (in the emulators) and it seems fine when using the demo page of the inputmask.

Can you reproduce it in an android emulator? or only on a real device?

best regards, 

Robin

**Comment by daanporon** [comment]
@RobinHerbots we couldn't reproduce this in the emulators only on the real devices. That's why i believe it has something to do with the performance of the device. If you have a device with poor performance maybe the setTimeout of 5ms is to long ... if you start typing before this timeout is triggered you'll loose a part of your input.

We are using version 2.2.28 of the input masks, but i see there is a new version availeble. Could it be that updating it will fix this?

We where experiencing it with all masks i believe, @roderik can you verify this?
It's possible to test them here:
https://www.bolero.be/nl/word-client/uw-gegevens (in dutch)
https://www.bolero.be/fr/devenez-client/vos-donnees (in french)
We always tested it on the birthday field (Geboortedatum, Date de naissance)

kind regards,
Daan

**Comment by RobinHerbots** [comment]
@daanporon, 

The new version will not fix the android issue but there is an important bug fixed, when you apply the same mask on multiple inputs at the same time.

BTW, Bolero isn't that an application from Acerta?  

best regards, 

Robin

**Comment by daanporon** [comment]
Ok, thanks i will update the version we are currently using.

Bolero is actually an application from KBC. You are also from Belgium?

Kind regards,
Daan

**Comment by RobinHerbots** [comment]
Yes I'm from belgium.  

To solve the android issue.  Maybe KBC can provide a real android device for a week or 2 so that I can test the inputmask on a real device.  Can this be arranged?  

Robin

**Comment by RobinHerbots** [comment]
@daanporon , @Dalmadog ,

Can you send me the useragent string of your android devices? 

Robin

**Comment by RobinHerbots** [comment]
see issue #186
